### DISCUSSION TITLE
NXP iMXRT106x generic FlexSPI driver support?

### DISCUSSION THREAD

**Discussion by asteriskSF** [zephyrproject-rtos/zephyr#71563] [Q&A] [answered]
@DerekSnell Does the flexspi driver in zephyr provide support for access to non-flash devices?  Do I need to develop my own driver to provide a generic SPI bus driver inteface?  

I've been told there is some McuXpresso SDK example that demonstrates using FlexSPI for generic SPI access, if this is possible, it would be great if NXP provided drivers to support SPI operation above 30 MHz to support data intensive peripherals.... unless there are significant impediments that maybe I should know about before embarking on my own effort?

**Comment by asteriskSF** [comment, marked as answer]
For anyone else interested, the FlexSPI programmable sequence engine can only read or write from the RX/TX FIFO's and therefore cannot implement full duplex SPI transceive operation.  Although not suitable for a full featured SPI implementation, it may be able to work as long as only one SPI buffer (rx or tx) is needed in each spi exchange.

**Reply by v-olin** [reply]
Is there any update on the question whether NXP would provide drivers for generic SPI access? I'm faced with a similar problem for an iMXRT1170 where I have to use a FlexSPI bus for generic SPI communication. You mentioned that you would embark on writing your own driver, did you get anywhere with it and if you finished i,t did it work well in your case?

**Comment by asteriskSF** [comment]
We were able to get approval to operate the generic SPI at a higher
frequency and did not need to implement generic SPI on the flex SPI.  From
what I recall it did not seem like the FlexSPI would be fully capable of
full-duplex bidirectional (at the same time) SPI, but it might have been
possible to achieve half-duplex SPI operations.

**Comment by DerekSnell** [comment]
Hi @asteriskSF and @v-olin ,
The FlexSPI peripheral is intended as a memory-mapped interface to external memories, including eXecute-In-Place (XIP) flash memories.  The FlexSPI is extremely flexible and programmable.  Likely it can be configured as a standard 1-bit SPI interface to non-memory devices.  However, I am not aware that NXP provides an example for that SPI use-case.  And today, NXP does not have any plans to enable Zephyr's SPI driver for the FlexSPI peripheral.

Best regards
